### PR TITLE
Added support for left- and right- versions of modifier keys.

### LINKED ISSUE #3347: Add Modifier keys left/right support
The thing is, in Godot the modifier keys **SHIFT, CTRL, ALT/OPTION and WINDOWS** keys for
both the left and right side are considered as the same key. Pressing left CTRL should be different
than pressing the right CTRL. Same for the other modifier keys. In Godot, left and right modifier keys share the same scancode.

From my experience with DirectInput in DirectX, it can be done. although Godot does not use DI.

Also, provide an option to disable propagating the **WINDOWS** key event to the OS, when Godot window is focused. Again, it can be done in DirectInput but don't know how Godot deals with Input. Currently, when I set up an action to use the windows key in game, when pressing it the start menu pops up.

### PR DESCRIPTION
**Pull request by EricEzaM**
Adds the ability to use Left/Right Shift, Control, Alt and Meta as key bindings.
These keys being used as modifiers (e.g. Shift+Ins, Ctrl+T) is unchanged, and either left or right will work.

References used:
https://docs.microsoft.com/en-us/windows/win32/inputdev/virtual-key-codes
https://www.win.tue.nl/~aeb/linux/kbd/scancodes-1.html

**Will close issues:**
Closes #3347.

### CODE REVIEW

**Comment by EricEzaM** [comment]
@akien-mga @reduz 

How do we allow for users to set a key to be "Any Shift/Control/Alt", so it doesn't matter which of left/right is pressed? Do we want this functionality? Sometimes they may just want alt/shift/ctrl to be pressed generically, rather than a specific side.

This might also go for editor shortcuts where we currently use KEY_CONTROL for things like going to definition in the Script editor... this might have to be changed to KEY_LEFT_CONTROL unless there is a way to make KEY_CONTROL represent any control key.

Please also note that since Mac does not differentiate between Left and Right Command, I have kept "Meta" as one key and have not split it into Left/Right variants.

**Comment by EricEzaM** [comment]
#3347

**Comment by EricEzaM** [comment]
* I added `bool Input::is_modifier_pressed(int p_keycode_mask)` where you can pass a mask like `KEY_MASK_CMD` or `KEY_MASK_SHIFT` and it will return whether one of the keys that make up that mask is pressed.
* I converted all code to use the above method. A **lot** of methods used `is_key_pressed(KEY_CONTROL)` which is wrong because we want it to act as Command on Apple and Control on others - which this would not do (would be Control on Apple too). Now, with `is_modifier_pressed(KEY_MASK_CMD)` it will correctly check for Command on Apple and Left/Right Control on Windows/Others.
* I removed the KEY_COMMAND definition as it was no longer compatible with the new Left/Right definitions. Also it was not used anywhere in the codebase, and was also not bound to scripting languages...
* Unfortunately, it seems now there is no easy cross-platform way to say `if Command pressed on Apple, or control pressed on Windows/Other`. We could potentially add the code below... but it is a bit misleading.
```
#ifdef APPLE_STYLE_KEYS
  KEY_COMMAND = KEY_META
#else
  KEY_COMMAND = KEY_LEFT_CONTROL
#endif
```
* Another option perhaps... faking left/right command?
#ifdef APPLE_STYLE_KEYS
  KEY_LEFT_COMMAND = KEY_META
  KEY_RIGHT_COMMAND = KEY_META
#else
  KEY_LEFT_COMMAND = KEY_LEFT_CONTROL
  KEY_RIGHT_COMMAND = KEY_RIGHT_CONTROL
#endif
```
* I had the idea which could potentially solve both of these, by abstracting the problem away from the `KeyList`. The key masks work perfectly fine since we do not discern between left/right for the modifier keys when they are being pressed. This is useful, like how I added `Input::is_modifier_pressed(KEY_MASK_CMD)`, but there is still the case of handling InputEventKey: `p_key->get_keycode() == KEY_LEFT_CONTROL || p_key->get_keycode() KEY_RIGHT_CONTROL`, which is still used quite a bit in the codebase. What if we added a method to InputEventKey which would check if the key pressed is a modifier key specifically. `bool InputEventKey::is_modifier_pressed(int p_keycode_mask)`, and it would use the same logic as the `Input` version. This way, there is an easy cross-platform way to know if "Command/Control" is pressed, since you could just do `p_key->is_modifier_pressed(KEY_MASK_CMD)`. This would test if the `keycode` of the event is KEY_META on Apple, or KEY_LEFT_CONTROL || KEY_RIGHT_CONTROL on Windows/Others.

**Comment by EricEzaM** [comment]
Ok, I implemented my idea in the most recent commit. I actually quite like it and think it is a pretty clean way of checking "is any shift key pressed", etc. Interested in feedback.

Look thru the commits one at a time and things should make more sense :)

**Comment by EricEzaM** [comment]
Took it out of draft and rebased... I can't test on Mac/Linux/others so if someone could vet the changes I made there it would be much appreciated.

File: platform/osx/display_server_osx.mm [hunk 1]
```diff
@@ -1118,14 +1118,14 @@ static int translateKey(unsigned int key) {
 		/* 34 */ KEY_UNKNOWN,
 		/* 35 */ KEY_ESCAPE,
 		/* 36 */ KEY_META,
-		/* 37 */ KEY_META,
-		/* 38 */ KEY_SHIFT,
+		/* 37 */ KEY_META, // Command/Meta on Mac apprears to only have one keycode: 55 (0x37), so left/right cannot be distinguished
```
Comment (bruvzg) [inline platform/osx/display_server_osx.mm:1121]: Left Meta is `55 (0x37)`, Right Meta is `54 (0x36)`.

Comment (EricEzaM) [reply platform/osx/display_server_osx.mm:1121]: Source?

Comment (EricEzaM) [reply platform/osx/display_server_osx.mm:1121]: I can't find any solid sources for this... A lot of mappings I have found just have 0x36 blank. But I will add it as you say anyway, I have found a couple of places that agree with that layout but like I said nothing solid.

Comment (bruvzg) [reply platform/osx/display_server_osx.mm:1121]: Right command on my Logitech MX Keys keyboard do produce 0x36 code on Big Sur, but maybe it's not true for some other keyboards. And it's listed in macOS SDK `/System/Library/Frameworks/Carbon.framework/Versions/A/Frameworks/HIToolbox.framework/Versions/A/Headers/Events.h` as `kVK_RightCommand = 0x36`.

Comment (EricEzaM) [reply platform/osx/display_server_osx.mm:1121]: Ah great, thanks!

**Comment by EricEzaM** [comment]
@reduz There is some decent support in the below issues for this functionality. 

https://github.com/godotengine/godot/issues/3347
https://github.com/godotengine/godot-proposals/issues/1387

I don't mind either way, at the time it was something that had been an outstanding issue for a while, and was one possible way of resolving https://github.com/godotengine/godot/issues/6851

I am not sure if @bruvzg was in the meeting and if he has any opinions on it (as he resolved the above issue in one of his PRs)

**Comment by Kehom** [comment]
> It would be nice to understand better which kind of use cases make this a necessary feature.

@reduz well, I don't know if this would qualify as "necessary feature", but pinball games typically use right and left shift keys to activate the respective flippers. Sometimes right and left Ctrl are then used to perform the "nudge" to simulate a "hit" in the machine from the respective side. While it would be possible to use different keys, I don't think it would be ideal to shift away from something that is somewhat "established".

**Comment by reduz** [comment]
@Kehom I know, the problem is the added complexity this adds. 

@EricEzaM Maybe one solution would be that we can have both something like
KEY_SHIFT, KEY_LEFT_SHIFT and KET_RIGHT_SHIFT and when you press left shift, you get _both_ of the key events sent. 

It would probably need an option in the action settings and shortcut setting when recording them that you want to discern between  using the modifier side or not.

**Comment by YuriSizov** [comment]
@bruvzg Can you be a tie-breaker and weigh in on the proposed options?

**Comment by bruvzg** [comment]
Not against it, but I do not think it's many use cases. Not every keyboard have both right and left key options, so it's definitly should have a way to map any of laft/right keys not just a specific one.

I would probably add an extra `is_left` and `is_right` properties to the `Key` event, instead of adding new modifier masks for each key. These properties can be normally ignored when matching actions, and checked manually if needed.

**Comment by Riteo** [comment]
@bruvzg that sounds like it could cause some confusion to some people as they might expect that it'd denote the physical position of the key (for example f is left and j is right).

**Comment by bruvzg** [comment]
> that sounds like it could cause some confusion to some people as they might expect that it'd denote the physical position of the key (for example f is left and j is right).

I guess nothing prevents us from actually setting l/r flags for all keys it this way (not sure why anyone would want it, bit it won't conflict with anything).

**Comment by EricEzaM** [comment]
At this stage I will not be continuing with this pull request. Anyone else is welcome to pick up the scraps if they are motivated to do so.

**Comment by Calinou** [comment]
As reduz mentioned above, I think it needs to be reimplemented this way to be considered:

> Maybe one solution would be that we can have both something like
KEY_SHIFT, KEY_LEFT_SHIFT and KET_RIGHT_SHIFT and when you press left shift, you get both of the key events sent.

**Comment by SchwingSK** [comment]
This solutions suits me, as I don't need the modifier status, and even if I needed it, I could implement it in my project, This means opening a new pull request, simpler than this one, I suppose ?

**Pull request closed without merging**